File: common/src/mobileMain/kotlin/jp/gree/techcon/common/viewstate/SessionDetailViewModel.kt
package jp.gree.techcon.common.viewstate

import jp.gree.techcon.common.AppState
import jp.gree.techcon.common.CommonModule
import jp.gree.techcon.common.logDebug
import jp.gree.techcon.common.model.Session
import jp.gree.techcon.common.usecase.SessionDetailUseCase
import kotlinx.coroutines.flow.Flow
import org.kodein.di.Kodein
import org.kodein.di.KodeinAware
import org.kodein.di.erased.instance

class SessionDetailViewModel(sessionId: Long): KodeinAware {
    override val kodein: Kodein by lazy { CommonModule.kodein }
    private val useCase: SessionDetailUseCase by kodein.instance()
    private val appState: AppState by kodein.instance()

    val session: Flow<Session>

    init {
        session = useCase.get(sessionId)
        // TODO: remove this
        logDebug("${appState.token}")
    }
}

File: iosApp/iosApp/ViewModel/SessionDetailViewModel.swift
//
//  SessionDetailViewModel.swift
//  GREE TechCon
//
//  Created by Yuto Yazaki on 2020/02/29.
//

import common

class SessionDetailViewModel: ObservableObject {
  @Published var session: Session?
  private let sessionId: Int
  private let sessionDetailService: SessionDetailService

  init(sessionId: Int, sessionDetailService: SessionDetailService = SessionDetailService()) {
    self.sessionId = sessionId
    self.sessionDetailService = sessionDetailService
    get()
  }

  func get() {
    sessionDetailService.get(sessionId: Int64(sessionId)).watch { [weak self] session in
      self?.session = session
    }
  }
}
